scipy/stats/_wilcoxon.py

```diff
@@ -179,9 +178,19 @@ def _wilcoxon_statistic(d, zero_method='wilcox'):
     se -= tie_correct/2
     se = np.sqrt(se / 24)
 
-    z = (r_plus - mn) / se
+    # se = 0 means that no non-zero values are left in d. usually this
+    # will be detected at the input validation stage (if method is asymptotic).
+    # handling this case here has two reasons:
+    # - if method="auto", the switch to asymptotic might only happen after
+    #   the statistic is calculated
+    # - for method != "asymptotic", avoid division by zero warning
+    #   (z is not needed anyways)
+    if np.any(se) == 0:
```

Comment:
this might need to change depending on what inputs we allow if ndim > 1

Reply:
I am thinking to add `output_z` as a keyword to the function that computes the statistic. It don't like the fact that you get a warning due to a calculation that is not needed unless `output_z = True`. So essentially the code would be:

```
if output_z:
     z = (r_plus - mn) / se
else:
   z = np.nan # no need to consider the dimension of se since the value is unused
```

wdyt?

Reply:
It needs to be calculated whenever the asymptotic method is used, and whether it is returned is already controlled by using `method='asymototic'`. SoI don't see a need for an additional keyword. 

I am not bothered by a warning one might get in an exceptional case. How did this come up - do you observe the warning frequently, happen to encounter it in the thousands of simulations you did with many ties, or did you see division by a variable in the code? If no user has complained about this in the past, I think there are other things we can do that will have greater impact. 

I'd like to keep the PR's scope to the four improvements, since it's complicated enough with those. If you consider this important enough to come back to separately, I'd be happy to discuss it then. Does that sound ok?

Reply:
the calculation happens within `_wilcoxon_statistic(d, zero_method='wilcox')`, so the method is not a keyword

the reason why I consider the warning relevant for this PR is as follows: I noted that we raise a value error if all entries are zero in a 1D array in input validation. since we plan to switch to asymptotic in some cases _after_ the validation step, I had to repeat this check. And then I noted that the behavior is not consistent with n-d arrays. I think we agree on the point that we can skip the additional validation and just return NaN and a warning. However, my view is that we should only show the warning if it is relevant (for example, if method="auto" and the sample size is 10, we switch to PermutationMethod and z is not needed/used, and it works fine with d = 0, so we should not warn). so for me, this change goes together with the change of the input validation. I hope it clarifies.

Reply:
I misunderstood when you wrote:

> I am thinking to add output_z as a keyword to the function that computes the statistic

I didn't notice that you wrote that you would add it to the private function, not the public function. I thought you were proposing a new public argument.

If the change is very simple, OK.